File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    bool check(vector<int>& arr, int m, int k, int mid) {
        int f = 0, cnt = 0;
        for (int i = 0; i < arr.size(); i++) {
            if (arr[i] <= mid) {
                cnt++;
                if (cnt == k) {
                    cnt = 0;
                    f++;
                }
            } else if (f >= m)
                return true;
            else if (arr[i] > mid) {
                cnt = 0;
            }
        }

        if (f < m)
            return false;
        return true;
    }
    int minDays(vector<int>& arr, int m, int k) {
        long long b = m,f = k;
        if (arr.size() < (b * f))
            return -1;
        int ans = 0, mid = 0;
        int st = INT_MAX, end = INT_MIN;
        for (int i = 0; i < arr.size(); i++) {
            st = min(st, arr[i]);
            end = max(end, arr[i]);
        }

        while (st <= end) {
            mid = st + (end - st) / 2;
            if (check(arr, m, k, mid)) {
                ans = mid;
                end = mid - 1;
            } else {
                st = mid + 1;
            }
        }
        return ans;
    }
};

File: 1605-minimum-number-of-days-to-make-m-bouquets/1605-minimum-number-of-days-to-make-m-bouquets.java
class Solution {
 boolean check(int[] bloomDay,int m,int k,int mid){
   int boq =0,flower=0;
    for(int it : bloomDay){
      if(it <= mid ){
        flower++;
        if(flower == k){
            boq++;
            flower = 0;
        }
    
      }
      else{
        flower = 0;
      }
    
    }
   return boq >=m;
}
    public int minDays(int[] bloomDay, int m, int k) {
        if(bloomDay.length < m*k) return -1;
        int st = Integer.MAX_VALUE,end = Integer.MIN_VALUE,mid=0,ans=-1;
         for(int it : bloomDay){
             st = Math.min(st,it);
             end = Math.max(end,it);
         }

         while(st<=end){
            mid = st + (end-st)/2;

            if(check(bloomDay,m,k,mid)){
              ans = mid;
              end = mid-1;
            }
            else{
                st = mid+1;
            }
         }
         return ans;
    }
}